### PR TITLE
Fix respecting font bounds

### PR DESCRIPTION
**Pull request by kuba2k2**
Yesterday, I have failed multiple times trying to evenly position an icon within a circle, just to find out there is an option to `respectFontBounds`, which is disabled by default. Enabling that manually on each of dozens `IconicsDrawable`s in my app would be a nightmare, which even turns to impossible for views added in XML (no such attribute).
This PR adds a global, default option (in the `Iconics` singleton) to enable `respectFontBounds` on every drawable/XML view.

Apparently, while the option was present, it had some issues. Sure, it did not scale the icons up (such as `gmd_arrow_drop_down`), but somehow the padding was still cropped out of all icons having uneven left-right paddings.
Pictures
(the right part comes from the Layout Inspector)

The option in the PR fixes this behavior (to truly fit the whole glyph in the canvas) and allows the user to add icon padding while still preserving the font's built-in padding.

The sample activity is updated to include a checkbox to toggle this option (having effect only in the large popup). The icon in each item has this always enabled, let me know if I should also change this.

### CODE REVIEW

File: app/src/main/java/com/mikepenz/iconics/sample/IconsFragment.kt [hunk 1]
```diff
@@ -37,22 +37,7 @@ import com.mikepenz.iconics.IconicsDrawable
...
-import com.mikepenz.iconics.utils.icon
-import com.mikepenz.iconics.utils.paddingDp
-import com.mikepenz.iconics.utils.roundedCorners
-import com.mikepenz.iconics.utils.roundedCornersDp
-import com.mikepenz.iconics.utils.shadowColor
-import com.mikepenz.iconics.utils.shadowDx
-import com.mikepenz.iconics.utils.shadowDy
-import com.mikepenz.iconics.utils.shadowRadius
-import com.mikepenz.iconics.utils.sizeDp
+import com.mikepenz.iconics.utils.*
```
Comment (zTrap) [inline app/src/main/java/com/mikepenz/iconics/sample/IconsFragment.kt:40]: Don't use start imports

File: iconics-core/src/main/java/com/mikepenz/iconics/IconicsDrawable.kt [hunk 2]
```diff
@@ -519,34 +521,39 @@ open class IconicsDrawable internal constructor() : WrappedDrawable() {
...
+        var textSize = paddingBounds.height().toFloat()
+        iconBrush.paint.textSize = textSize
+        iconBrush.paint.getTextPath(textValue, 0, textValue.length, 0f, 0f, path)
         path.computeBounds(pathBounds, true)
 
-        if (!respectFontBounds) {
+        if (respectFontBounds) {
+            path.offset(viewBounds.exactCenterX(), paddingBounds.top + textSize - iconBrush.paint.fontMetrics.descent)
+        }
+        else {
```
Comment (zTrap) [inline iconics-core/src/main/java/com/mikepenz/iconics/IconicsDrawable.kt:535]: else must be at the same line with if's `}`

File: iconics-core/src/main/java/com/mikepenz/iconics/IconicsDrawable.kt [hunk 3]
```diff
@@ -519,34 +521,39 @@ open class IconicsDrawable internal constructor() : WrappedDrawable() {
...
-
-        val startY = viewBounds.centerY() - pathBounds.height() / 2
-        val offsetY = startY - pathBounds.top
-
-        path.offset(offsetX + iconOffsetXPx, offsetY + iconOffsetYPx)
+    private fun offsetIcon() {
+        if (respectFontBounds) {
+            path.offset(iconOffsetXPx.toFloat(), iconOffsetYPx.toFloat())
+        }
+        else {
```
Comment (zTrap) [inline iconics-core/src/main/java/com/mikepenz/iconics/IconicsDrawable.kt:552]: else must be at the same line with if's }

**Comment by mikepenz** [comment]
thank you so much @kuba2k2 for this PR

**Review by zTrap — APPROVED** [review]

**Comment by kuba2k2** [comment]
Hey, does this PR require any further changes, or would it be ok for you to merge/release it?

**Comment by mikepenz** [comment]
@kuba2k2 the PR itself should be fine. I haven't had time to review it in depth yet, which I need to do before merge, want to ensure no behaviour change for existing users

**Comment by mikepenz** [comment]
Thank you so much. I'll include this in a beta first to see if the adjustment have any un-anticipated side effects

**Comment by kuba2k2** [comment]
Thanks, I can now start adapting my app to usage with `respectFontBoundsDefault` as true. As you probably already know, with this option set, all icons would be smaller - because of the font's inbuilt padding. Although I did not notice any changes between 5.2.8 and my PR (when leaving the option as default - false).

**Comment by mikepenz** [comment]
@kuba2k2 thank you for opening the PR :) 

Yes indeed. I checked and compared. things looked good, just wanting to play it double safe to not break somebodies integration un-intentional. having it in beta for a bit will hopefully give people enough time to pre-test and report if they see something strange.

**Comment by kuba2k2** [comment]
Hey, will/is the `5.3.0-b01` published to Maven Central? I can't find nor resolve it in Gradle, are you waiting for some kind of a review or acceptance (never published to Maven Central) or haven't uploaded it yet? Not that I'm hurrying you up, just want to know if there's something wrong on my side.

**Comment by mikepenz** [comment]
@kuba2k2 thanks so much for pinging. totally forgot to do the staging procedure. so after publishing the staging repo needs to be closed and then released. 

after publishing to bintray (where this is not required) for a few years I totally forgot. 

Did this now. should be up in 30-60 min. 

all should be good on your end

**Comment by kuba2k2** [comment]
Huh, now I see this change may actually be problematic (if the default value gets enabled). For example, consider the MaterialDrawer's primary item, which sets a 1dp padding to a 24dp icon, leaving 22dp for the glyph, which is now only 18dp for the icon itself. As per the Material Design guidelines, the icon should be 24dp for the glyph. Do you know why is this 1dp padding added in MaterialDrawer? Is it maybe because the paddings were cropped by Iconics at the time? Should we maybe update MD to respect font bounds and remove the 1dp padding? I'm asking, because the icons now look so... tiny..

**Comment by mikepenz** [comment]
The `1dp` padding was added back in the days to ensure the right size of the icons to match the exact spec. Especially with fonts which do not have the proper paddings this was super important. 

The community material icon font luckily includes them precisely. 

I wonder though what the right decision here is. ultimately it may break projects silently (by the paddings being wrong) 

would for your usecase maybe an extended `IconicsDrawable` or extension function work which will automatically enable the respect of the bounds wherever you use them? (so you would not need the global flag?)

**Comment by kuba2k2** [comment]
Not really, I think, because I'm using the IconicsImageView in some places, and even then I want to respect the bounds in most cases. I will probably update the MD's ImageHolder to just ignore the 1dp padding, as I'm using a custom holder anyway.

As for the 1dp padding maybe you're right it should remain there. Or maybe there should also be an option to enable/disable it, according to which font the developer uses. So if the app uses for example FontAwesome (in which the paddings are totally broken, btw) the developer would enable the inbuilt padding, but I, using CMD, would disable it easily.

**Pull request merged**